src/main/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListBatchValidator.java

```diff
@@ -0,0 +1,74 @@
+import com.commercetools.sync.shoppinglists.ShoppingListSyncOptions;
+import io.sphere.sdk.shoppinglists.ShoppingListDraft;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+public class ShoppingListBatchValidator
+    extends BaseBatchValidator<ShoppingListDraft, ShoppingListSyncOptions, ShoppingListSyncStatistics> {
+
+    static final String SHOPPING_LIST_DRAFT_KEY_NOT_SET = "ShoppingListDraft with name: %s doesn't have a key. "
+        + "Please make sure all shopping list drafts have keys.";
+    static final String SHOPPING_LIST_DRAFT_IS_NULL = "ShoppingListDraft is null.";
+
+    public ShoppingListBatchValidator(@Nonnull final ShoppingListSyncOptions syncOptions,
+                                      @Nonnull final ShoppingListSyncStatistics syncStatistics) {
+        super(syncOptions, syncStatistics);
+    }
+
+    /**
+     * Given the {@link List}&lt;{@link ShoppingListDraft}&gt; of drafts this method attempts to validate
+     * drafts and collect referenced type keys from the draft and return an {@link ImmutablePair}&lt;{@link Set}&lt;
+     * {@link ShoppingListDraft}&gt;,{@link Set}&lt;
+     * {@link String}&gt;&gt; which contains the {@link Set} of valid drafts and referenced type keys.
+     *
+     * <p>A valid shopping list draft is one which satisfies the following conditions:
+     * <ol>
+     * <li>It is not null</li>
+     * <li>It has a key which is not blank (null/empty)</li>
+     * </ol>
+     *
+     * @param shoppingListDrafts the shopping list drafts to validate and collect referenced type keys.
+     * @return {@link ImmutablePair}&lt;{@link Set}&lt;{@link ShoppingListDraft}&gt;,
+     *      {@link Set}&lt;{@link String}&gt;&gt; which contains the {@link Set} of valid drafts and
+     *      referenced type keys.
+     */
+    @Override
+    public ImmutablePair<Set<ShoppingListDraft>, Set<String>> validateAndCollectReferencedKeys(
+        @Nonnull final List<ShoppingListDraft> shoppingListDrafts) {
+        final Set<String> typeKeys = new HashSet<>();
+        final Set<ShoppingListDraft> validDrafts = shoppingListDrafts
+            .stream()
+            .filter(this::isValidShoppingListDraft)
+            .peek(shoppingListDraft ->
+                collectReferencedKeyFromCustomFieldsDraft(shoppingListDraft.getCustom(), typeKeys::add))
+            .collect(Collectors.toSet());
+
+        return ImmutablePair.of(validDrafts, typeKeys);
+    }
+
+    private boolean isValidShoppingListDraft(
+        @Nullable final ShoppingListDraft shoppingListDraft) {
+
+        if (shoppingListDraft == null) {
+            handleError(SHOPPING_LIST_DRAFT_IS_NULL);
+        } else if (isBlank(shoppingListDraft.getKey())) {
```

Comment:
ok, this is a LocalizedString, so a nullcheck is enough or should i check the Map is empty too?

Reply:
my message is clear for the validation purpose, yes you need to validate/check otherwise it will be an unnecessary call to API and it will return 400, in validation methods we are trying to avoid this.

I have checked the SDK part for you, so from that, you could create an empty name, but not a duplicate name.

```` java
   @Test
    public void testShoppingListDraftName () {
        final ShoppingListDraftBuilder emptyName =
            ShoppingListDraftBuilder.of(LocalizedString.of());

        assertThat(emptyName).extracting(ShoppingListDraftBuilder::getName)
                             .isNotNull();

        assertThatThrownBy(() -> ShoppingListDraftBuilder.of(
            LocalizedString.of(Locale.ENGLISH, "a", Locale.ENGLISH, "b")))
            .hasCauseInstanceOf(IllegalArgumentException.class)
            .hasMessage("Duplicate keys (en) for map creation.");
    }
````
I tested with an empty name platform will throw:

````
{
    "statusCode": 400,
    "message": "Request body does not contain valid JSON.",
    "errors": [
        {
            "code": "InvalidJsonInput",
            "message": "Request body does not contain valid JSON.",
            "detailedErrorMessage": "name: Values of LocalizedString must not be empty."
        }
    ]
}
````

Reply:
Thanks for the investigation!

Reply:
23697a7

Reply:
I have more comments related, will you update, I also wrote to slack: https://github.com/commercetools/commercetools-sync-java/pull/595#pullrequestreview-508137025

Reply:
396ce83